x/concentrated-liquidity/internal/swapstrategy/one_for_zero.go

```diff
@@ -78,17 +78,124 @@ func (s oneForZeroStrategy) ComputeSwapStep(sqrtPriceCurrent, sqrtPriceNextTick,
 
 	// Handle fees.
 	// Note that fee is always charged on the amount in.
-	var feeChargeTotal sdk.Dec
-	if s.isOutGivenIn {
-		// amountOne is amountIn.
-		feeChargeTotal = computeFeeChargePerSwapStepOutGivenIn(sqrtPriceCurrent, hasReachedTarget, amountOne, amountRemaining, s.swapFee)
+
+	// amountZero is amountIn.
+	// TODO: multiplication with rounding up at precision end.
+	feeChargeTotal := amountZero.Mul(s.swapFee).Quo(sdk.OneDec().Sub(s.swapFee))
+
+	return sqrtPriceNext, amountOne, amountZero, feeChargeTotal
+}
+
+// ComputeSwapStepOutGivenIn calculates the next sqrt price, the amount of token in consumed, the amount out to return to the user, and total fee charge on token in.
+// Parameters:
+//   - sqrtPriceCurrent is the current sqrt price.
+//   - sqrtPriceTarget is the target sqrt price computed with GetSqrtTargetPrice(). It must be one of:
+//     1. Next tick sqrt price.
+//     2. Sqrt price limit representing price impact protection.
+//   - liquidity is the amount of liquidity between the sqrt price current and sqrt price target.
+//   - amountOneRemainingIn is the amount of token one in remaining to be swapped. This amount is fully consumed
+//     if sqrt price target is not reached. In that case, the returned amountOne is the amount remaining given.
+//     Otherwise, the returned amountOneIn will be smaller than amountOneRemainingIn given.
+//
+// Returns:
+//   - sqrtPriceNext is the next sqrt price. It equals sqrt price target if target is reached. Otherwise, it is in-between sqrt price current and target.
+//   - amountOneIn is the amount of token in consumed. It equals amountRemainingIn if target is reached. Otherwise, it is less than amountOneRemainingIn.
+//   - amountZeroOut the amount of token out computed. It is the amount of token out to return to the user.
+//   - feeChargeTotal is the total fee charge. The fee is charged on the amount of token in.
+//
+// OneForZero details:
+// - oneForZeroStrategy assumes moving to the right of the current square root price.
+func (s oneForZeroStrategy) ComputeSwapStepOutGivenIn(sqrtPriceCurrent, sqrtPriceTarget, liquidity, amountOneInRemaining sdk.Dec) (sdk.Dec, sdk.Dec, sdk.Dec, sdk.Dec) {
+	// Estimate the amount of token one needed until the target sqrt price is reached.
+	amountOneIn := math.CalcAmount1Delta(liquidity, sqrtPriceTarget, sqrtPriceCurrent, true) // N.B.: if this is false, causes infinite loop
+
+	// Calculate sqrtPriceNext on the amount of token remaining after fee.
+	amountOneInRemainingLessFee := amountOneInRemaining.Mul(sdk.OneDec().Sub(s.swapFee))
+
+	var sqrtPriceNext sdk.Dec
+	// If have more of the amount remaining after fee than estimated until target,
+	// bound the next sqrtPriceNext by the target sqrt price.
+	if amountOneInRemainingLessFee.GTE(amountOneIn) {
```

Comment:
What is the scenario where `amountOneInRemainingLessFee` would be _greater than_ `amountOneIn`?